hq_superset/utils.py

```diff
@@ -189,3 +226,53 @@ def array_is_falsy(array_values):
         return []
 
     return array_values
+
+
+def js_to_py_datetime(jsdt, preserve_tz=True):
+    """
+    JavaScript UTC datetimes end in "Z". In Python < 3.11,
+    ``datetime.isoformat()`` doesn't like it, and raises
+    "ValueError: Invalid isoformat string"
+
+    >>> jsdt = '2024-02-24T14:01:25.397469Z'
+    >>> js_to_py_datetime(jsdt)
+    datetime.datetime(2024, 2, 24, 14, 1, 25, 397469, tzinfo=datetime.timezone.utc)
+    >>> js_to_py_datetime(jsdt, preserve_tz=False)
+    datetime.datetime(2024, 2, 24, 14, 1, 25, 397469)
+
+    """
+    if preserve_tz:
+        if sys.version_info >= (3, 11):
+            return datetime.fromisoformat(jsdt)
+        pydt = jsdt.replace('Z', '+00:00')
+    else:
+        pydt = jsdt.replace('Z', '')
+    return datetime.fromisoformat(pydt)
+
+
+def cast_data_for_table(
+    data: list[dict[str, Any]],
+    table: TableClause,
+) -> Generator[dict[str, Any], None, None]:
+    """
+    Returns ``data`` with values cast in the correct data types for
+    the columns of ``table``.
+    """
+    cast_functions = {
+        # 'BIGINT': int,
+        # 'TEXT': str,
+        'TIMESTAMP': partial(js_to_py_datetime, preserve_tz=False),
+        # TODO: What else?
+    }
+
+    column_types = {c.name: str(c.type) for c in table.columns}
+    for row in data:
```

Comment:
nit: probably its just me, though when I see `row`, I assume its a list. But here, this is a dictionary. That got me.

Reply:
That's what the type hint on line 256 is for.

While "row" could be a list, or a tuple, or a dictionary, "data" could be absolutely anything! Any variable contains data, doesn't it? And "table" is also ambiguous -- SQLAlchemy documentation uses the variable name "table" to refer to two or three different types.

So type hints, _in this specific case_, are helpful. The alternative would be to rename "data" to something more accurate, but less friendly, like "list_of_column_value_pairs", but then you wouldn't get any validation from your IDE.

Fun exercise: Open this in your IDE. Ctrl-click on the "columns" property of "table". Magic, right? Now delete the type hint, and try again.

Reply:
(You kinda handed me my favorite drum, and I just went ahead and banged it. :joy: :speak_no_evil:  Sorry about that.)

Reply:
I kinda new what I was getting back when i asked that question, I knew you would find your way to typehints.